docs/design/services/s3/transfermanager/README.md

```diff
@@ -0,0 +1,66 @@
+# Project Tenets (unless you know better ones)
+
+1. Meeting customers in their problem space allows them to deliver value
+   quickly.
+2. Meeting customer expectations drives usability.
+3. Discoverability drives usage.
+
+# Introduction
+
+This project provides a much improved experience for S3 customers needing to
+easily perform uploads and downloads of objects to and from S3 by providing
+the S3 `TransferManager`, a high level library built on the S3 client.
+
+# Project Goals
+
+1. For the use cases it addresses, i.e. the transfer of objects to and from S3,
+   TransferManager is the preferred solution. It is easier and more intuitive
+   than using the S3 client. In the majority of situations, it is more
+   performant.
+1. TransferManager provides a truly asynchronous API that conforms to the norms
+   present in the rest of the SDK.
+1. TransferManager makes efficient use of system resources.
+1. TransferManager performs well in cold JVMs.
+1. TransferManager supplements rather than replaces the lower level S3 client.
+
+# Non Project Goals
+
+1. Ability to use the blocking, synchronous client.
```

Comment:
Not sure where you separate client vs http client but I think it should be a primary goal to support sync http clients like Apache. Each HTTP client has different levels of configuration and allowing customers to use their preferred one should be a requirement.

Reply:
I guess both as in the blocking service client and HTTP clients. I can't see supporting a sync HTTP client as a goal can coexist with the async/non-blocking, efficiency, and performance goals. We already saw a lot of this when we explored whether we could expose sync HTTP clients as async and vice versa.

To me, a big part of the original TransferManager was to provide the async/non-blocking experience so by focusing on using the async HTTP clients, we further enhance that experience.

If it's a matter of configuration I would think supporting a config present in Apache but not Netty would be easier to achieve than supporting Apache/a non blocking HTTP client in TransferManager.

Reply:
I think just limiting it to non-blocking clients is fine, as long as we have support for HTTP proxies in Netty. Sync clients and async abstractions like TransferManager don't mix well.

Reply:
So no shared connection pool if customers are using the sync s3 client? Which is made worse given that you have to use both the low-level client and transfer manager.

No lightweight JDK client?

No GAE support?

A worse debugging experience.

Reply:
In my view, TransferManager is an enhancement of the async client, where this is all true right now as well.

I think it's important that customers who can't use the async clients for the reasons given above are not left without an option, but I think that means we provide a similar, but separate library for non-async clients.